python/tvm/autotvm/tuner/model_based_basetuner.py

```diff
@@ -0,0 +1,179 @@
+    sa_temp: float or list of float
+        Temperature config of simulated annealing
+    sa_persistent: bool
+        Whether keep persistent states of SA points among different models
+    sa_parallel_size: int
+        Number of parallel Markov chains when doing parallel simulated annealing
+    """
+
+    def __init__(self, task, batch_size,
+                 sa_n_iter, sa_temp, sa_persistent, sa_parallel_size):
+        super(ModelBasedBaseTuner, self).__init__(task)
+
+        self.batch_size = batch_size
+        self.space_len = len(task.config_space)
+
+        # space
+        self.task = task
+        self.target = task.target
+        self.space = task.config_space
+        self.dims = [len(x) for x in self.space.space_map.values()]
+
+        # trial planning
+        self.trials = []
+        self.trial_pt = 0
+        self.visited = set()
+
+        # observed samples
+        self.xs = []
+        self.ys = []
+        self.flops_max = 0.0
+        self.y_max = 0
+
+        # simulated annealing for optimizing acquisition function
+        self.sa_n_iter = sa_n_iter
+        self.sa_temp = sa_temp
+        self.sa_persistent = sa_persistent
+        self.sa_parallel_size = min(sa_parallel_size, len(self.space))
+        self.sa_points = None
+        self.sa_trans_func = None
+        self.sa_eval_func = None
+
+        self.train_ct = 0
+
+    def next_batch(self, batch_size):
+        ret = []
+
+        counter = 0
+        while counter < batch_size:
+            if len(self.visited) >= len(self.space):
+                break
+
+            while self.trial_pt < len(self.trials):
+                index = self.trials[self.trial_pt]
+                if index not in self.visited:
+                    break
+                self.trial_pt += 1
+
+            if self.trial_pt >= len(self.trials):  # trial list is empty, choose randomly
+                index = np.random.randint(len(self.space))
+                while index in self.visited:
+                    index = np.random.randint(len(self.space))
+
+            ret.append(self.space.get(index))
+            self.visited.add(index)
+
+            counter += 1
+        return ret
+
+    def update(self, inputs, results):
+        for inp, res in zip(inputs, results):
+            index = inp.config.index
+            if res.error_no == 0:
+                self.xs.append(index)
+                flops = inp.task.flop / np.mean(res.costs)
+                self.flops_max = max(self.flops_max, flops)
+                self.ys.append(flops)
+            else:
+                self.xs.append(index)
+                self.ys.append(0)
+
+        self._update_model()
+
+    def _update_model(self):
+        """refit a new model if the tuner collects enough new samples"""
+        pass
+
+    def has_next(self):
+        return len(self.visited) < len(self.space)
+
+    def save_state(self, filename):
```

Comment:
save/load state are not pythonic

Instead, directly make them pickable. https://docs.python.org/2/library/pickle.html#object.__getstate__